### PR TITLE
Admin table snapshots api

### PR DESCRIPTION
**Pull request by jiren**
- Table snapshot create, get, list and delete, create a table from snapshot API
- GRPC mock test cases

### CODE REVIEW

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb [hunk 1]
```diff
@@ -227,6 +227,280 @@ def drop_row_range \
...
+        #   end
+        #
+        #   # List all snapshots from all clusters
+        #   client.snapshots.each do |snapshot|
+        #     # Process snapshot
+        #     p snapshot
+        #   end
+
+        def snapshots \
+            cluster_id = "-",
```
Comment (quartzmo) [inline google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: I don't see a test for the default `cluster_id` value of `"="`.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb [hunk 2]
```diff
@@ -227,6 +227,280 @@ def drop_row_range \
...
+        #   cluster, e.g., +mysnapshot+ of the form: +[_a-zA-Z0-9][-_.a-zA-Z0-9]*+
+        # @param description [String]
+        #   Description of the snapshot.
+        # @param ttl [Integer]
+        #   The amount of time in seconds that the new snapshot can stay active
+        #   after it is created. Once 'ttl' expires, the snapshot will get
+        #   deleted. The maximum amount of time a snapshot can stay active is
+        #   7 days. If 'ttl' is not specified, the default value of 24 hours
+        #   will be used.
+        # @param options [Google::Gax::CallOptions]
```
Comment (blowmage) [inline google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: We don't usually expose `Google::Gax::CallOptions` in the high level API.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb [hunk 3]
```diff
@@ -227,6 +227,280 @@ def drop_row_range \
...
+        #   Overrides the default settings for this call, e.g, timeout,
+        #   retries, etc.
+        # @return [Google::Longrunning::Operation]
+        # @raise [Google::Gax::GaxError] if the RPC is aborted.
```
Comment (blowmage) [inline google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: We have been wrapping `Google::Gax::GaxError` with a `Google::Cloud::Error`.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb [hunk 4]
```diff
@@ -227,6 +227,280 @@ def drop_row_range \
...
+        #   )
+
+        def create_snapshot \
+            table_id,
+            cluster_id,
+            snapshot_id,
+            description,
+            ttl: nil,
+            options: nil
+          ttl = Google::Protobuf::Duration.new(seconds: ttl) if ttl
```
Comment (blowmage) [inline google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: We use this helper method in some of the other gems:

```ruby
def number_to_duration number
  return nil if number.nil?

  Google::Protobuf::Duration.new \
    seconds: number.to_i,
    nanos: (number.remainder(1) * 1000000000).round
end
```

See [`Convert.number_to_duration`](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-pubsub/lib/google/cloud/pubsub/convert.rb#L42-L48)

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb (hunk 1 above)
Comment (jiren) [reply google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: I will add.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb (hunk 3 above)
Comment (jiren) [reply google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: I will make a change.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb (hunk 2 above)
Comment (jiren) [reply google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: Got it. I will update in all other APIs also. 

In some of the APIs call options require i.e In `drop_rows` BigTable server is responding with Error - increase timeout limit and try again.
I am planning to add an option for `timeout` for such particular API call.

File: google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb (hunk 4 above)
Comment (jiren) [reply google-cloud-bigtable/lib/google/cloud/bigtable/table_admin_client.rb]: I will take a reference from it.

**Comment by sduskis** [comment]
Snapshots are an alpha feature on the service.  This R should be considered a low priority.

**Comment by jiren** [comment]
@sduskis I am closing this pull request. All admin apis are covered in https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/2128

**Pull request closed without merging**